chia/protocols/farmer_protocol.py

```diff
@@ -16,6 +20,26 @@
 Note: When changing this file, also change protocol_message_types.py, and the protocol version in shared_protocol.py
 """
 
+@streamable
+@dataclass(frozen=True)
+class SPSubSlotSourceData(Streamable):
+    cc_sub_slot: ChallengeChainSubSlot
+    rc_sub_slot: RewardChainSubSlot
+
+
+@streamable
+@dataclass(frozen=True)
+class SPVDFSourceData(Streamable):
+    cc_vdf: ClassgroupElement
+    rc_vdf: ClassgroupElement
+
+
+@streamable
+@dataclass(frozen=True)
+class SignagePointSourceData(Streamable):
+    sub_slot_data: Optional[SPSubSlotSourceData]
+    vdf_data: Optional[SPVDFSourceData]
```

Comment:
this is not a strong opinion, but it seems like the existing protocol suffers from trying to over-generalize (by just sending an arbitrary number of un-typed hashes to be signed). Rather than doubling down on this generalization, I think there's an opportunity to simplify the protocol to more closely match what we actually use and need.

Rather than sending both the hashes (un-typed) and then a (relatively deep) structure with typed source-data, couldn't you just send the source data as optional fields? and whichever field is set indicates what we want to be signed. Maybe we can even make the hashes optional, and only send those when we use our harvester that doesn't verify what it's signing.

Reply:
I'm not an expert in python, but in C++ you would just have base class that every structure inherits from and then you can pass a list of pointers with the base type. The implementation can then check each instance what the actual type is if needed, but the hash function would be generic. There would not be any need to invent new types just for this purpose, or using a list of optional fields. If that is possible in python I have no idea...

Reply:
I opted for a more conservative approach when modifying the protocol since I was not aware as to the reasoning behind taking a generalized approach when signing the messages. If I have your blessing to make it more specific and explicit, I am all for it.

Reply:
I'm a proponent of "you-ain't-gonna-need-it". i.e. keep things as concrete (and simple) as possible, and refactor later if you need to.

Reply:
> in C++ you would just have base class that every structure inherits from and then you can pass a list of pointers with the base type

This type describes the network protocol, and our serialization is somewhat limited. We don't support a variant type for example. The alternative would be a series of optional fields, where one is set.